Configuring a Setup.hs script without cabal-doctest installed

This isn't an issue when using `cabal-doctest` so much as it is an issue when _not_ using `cabal-doctest`. Or, to put it it more clearly, we have several backwards-compatible `Setup` scripts which guard against the presence of `cabal-doctest` like so:

```haskell
#if MIN_VERSION_cabal_doctest(1,0,0)
-- This should be the code you use with Cabal-1.24 or later
import Distribution.Extra.Doctest ( defaultMainWithDoctests )
#else
-- Backported cabal-doctest code
-- In theory, this code should only ever be reachable with old versions of Cabal
#endif

main :: IO ()
main = defaultMainWithDoctests "doctests"
```

As I put in the comments above, there's a bit of an assumption that when you have a recent version of `Cabal` installed, you always go down one code path, and when you have an older `Cabal`, you go down the other, legacy code path. When _installing_ a package that uses this sort of script, this assumption should always hold, since your build tool will make sure you've installed `cabal-doctest` if you require it (i.e., if you're using a recent `Cabal`).

But https://github.com/ekmett/lens/issues/721 revealed that this assumption doesn't always hold when you simply _configure_ a package. Here's a simple way to illustrate the problem:

```
$ /opt/cabal/2.0/bin/cabal get distributive
Unpacking to distributive-0.5.2/
$ cd distributive-0.5.2/
$ /opt/cabal/2.0/bin/cabal sandbox init
$ /opt/cabal/2.0/bin/cabal configure -w /opt/ghc/8.2.1/bin/ghc
Resolving dependencies...
Warning: solver failed to find a solution:
Could not resolve dependencies:
trying: distributive-0.5.2:+tagged
unknown package: tagged (dependency of distributive-0.5.2:+tagged)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: distributive, distributive-0.5.2:flag,
taggedTrying configure anyway.
[1 of 1] Compiling Main             ( dist/setup/setup.hs, dist/setup/Main.o )

dist/setup/setup.hs:103:7: error:
    • Couldn't match expected type ‘IO ()’
                  with actual type ‘String -> IO ()’
    • In a stmt of a 'do' block:
        rewriteFile (testAutogenDir </> "Build_doctests.hs")
          $ unlines
              ["module Build_doctests where", "", "pkgs :: [String]",
               "pkgs = " ++ (show $ formatDeps $ testDeps libcfg suitecfg), ....]
      In the second argument of ‘($)’, namely
        ‘do let testAutogenDir = autogenComponentModulesDir lbi suitecfg
            createDirectoryIfMissingVerbose verbosity True testAutogenDir
            rewriteFile (testAutogenDir </> "Build_doctests.hs")
              $ unlines ["module Build_doctests where", "", ....]’
      In the expression:
        when (testName suite == fromString testsuiteName)
          $ do let testAutogenDir = autogenComponentModulesDir lbi suitecfg
               createDirectoryIfMissingVerbose verbosity True testAutogenDir
               rewriteFile (testAutogenDir </> "Build_doctests.hs")
                 $ unlines ["module Build_doctests where", "", ....]
    |
103 |       rewriteFile (testAutogenDir </> "Build_doctests.hs") $ unlines
    |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...

dist/setup/setup.hs:103:20: error:
    • Couldn't match type ‘[Char]’
                     with ‘Distribution.Verbosity.Verbosity’
      Expected type: Distribution.Verbosity.Verbosity
        Actual type: FilePath
    • In the first argument of ‘rewriteFile’, namely
        ‘(testAutogenDir </> "Build_doctests.hs")’
      In the expression:
        rewriteFile (testAutogenDir </> "Build_doctests.hs")
      In a stmt of a 'do' block:
        rewriteFile (testAutogenDir </> "Build_doctests.hs")
          $ unlines
              ["module Build_doctests where", "", "pkgs :: [String]",
               "pkgs = " ++ (show $ formatDeps $ testDeps libcfg suitecfg), ....]
    |
103 |       rewriteFile (testAutogenDir </> "Build_doctests.hs") $ unlines
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

dist/setup/setup.hs:168:18: error:
    • Couldn't match type ‘Distribution.Types.MungedPackageId.MungedPackageId’
                     with ‘Distribution.Types.PackageId.PackageIdentifier’
      Expected type: [(UnitId, PackageId)]
        Actual type: [(UnitId,
                       Distribution.Types.MungedPackageId.MungedPackageId)]
    • In the expression:
        nub $ componentPackageDeps xs ++ componentPackageDeps ys
      In an equation for ‘testDeps’:
          testDeps xs ys
            = nub $ componentPackageDeps xs ++ componentPackageDeps ys
    |
168 | testDeps xs ys = nub $ componentPackageDeps xs ++ componentPackageDeps ys
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Ack! Things went wrong here because we have a recent `Cabal` installed, but since we were in a fresh sandbox, we also didn't have `cabal-doctest` installed. Therefore we went down the wrong (legacy) codepath, and unsurprisingly, the more recent `Cabal` API doesn't work against the legacy code.

My question is: is there anything we can do here to lessen the surprise for users who might run into this situation? Obviously, we can't keep updating the legacy code in these `Setup` scripts forever to support new versions of `Cabal`—after all, we created `cabal-doctest` in the first place to avoid having to do that! Perhaps we should emit a large, scary GHC warning whenever you go down the legacy codepath with a recent `Cabal`? Or is there a more graceful way to handle this?

--- a/README.md
+++ b/README.md
@@ -8,7 +8,10 @@ A `Setup.hs` helper for running `doctests`.
 Example Usage
 =============
 
-See [https://github.com/phadej/cabal-doctest/tree/master/example] for an example package.
+See [https://github.com/phadej/cabal-doctest/tree/master/example] for an
+example package. (Note that the example requires `Cabal-1.24` or later, but
+you can relax this bound safely, although running doctests won't be supported
+on versions of `Cabal` older than 1.24.)
 
 To use this library in your `Setup.hs`, you should specify a `custom-setup`
 section in your `.cabal` file. For example:

--- a/example/Setup.hs
+++ b/example/Setup.hs
@@ -1,6 +1,33 @@
-module Main where
+{-# LANGUAGE CPP #-}
+{-# OPTIONS_GHC -Wall #-}
+module Main (main) where
 
-import Distribution.Extra.Doctest (defaultMainWithDoctests)
+#ifndef MIN_VERSION_cabal_doctest
+#define MIN_VERSION_cabal_doctest(x,y,z) 0
+#endif
 
+#if MIN_VERSION_cabal_doctest(1,0,0)
+
+import Distribution.Extra.Doctest ( defaultMainWithDoctests )
 main :: IO ()
 main = defaultMainWithDoctests "doctests"
+
+#else
+
+#ifdef MIN_VERSION_Cabal
+-- If the macro is defined, we have new cabal-install,
+-- but for some reason we don't have cabal-doctest in package-db
+--
+-- Probably we are running cabal sdist, when otherwise using new-build
+-- workflow
+#warning You are configuring this package without cabal-doctest installed. \
+         The doctests test-suite will not work as a result. \
+         To fix this, install cabal-doctest before configuring.
+#endif
+
+import Distribution.Simple
+
+main :: IO ()
+main = defaultMain
+
+#endif

